### ISSUE TITLE
Allow storing `format_args!` in a let binding

### ISSUE THREAD

**Issue by piegamesde** [rust-lang/rust#92698]
At the moment, we cannot store the result of a `format_args!` in a value:

```rust
// Okay, we can somehow store it in a variable
let foo = format_args!("{} foo {:?}", 1, 2);

// But once we try to use it, we get
// "error[E0716]: temporary value dropped while borrowed"
// "note: consider using a `let` binding to create a longer lived value"
// "note: this error originates in the macro `format_args`"
println!("{}", foo);

// Using it directly is okay though
println!("{}", format_args!("{} foo {:?}", 1, 2));

// We can get pretty creative within the "direct usage" constraints :D
let foo = match format_args!("{} foo {:?}", 1, 2) {
    foo => {println!("{}", foo);},
};

// But even if we manage to store the value in a struct, …
struct Foo<'a> {foo: std::fmt::Arguments<'a> }
let foo = Foo { foo: format_args!("{} foo {:?}", 1, 2) };

// … we still cannot make use of it outside of the scope it was created in
println!("{}", foo.foo); // <- error[E0716]: temporary value dropped while borrowed
```

The list of confused or annoyed users by this is rather long:

- #42253
- #11838
- #51905
- https://docs.rs/fern/0.6.0/fern/struct.FormatCallback.html
- https://stackoverflow.com/questions/56304313/cannot-use-format-args-due-to-temporary-value-is-freed-at-the-end-of-this-state
- https://users.rust-lang.org/t/using-format-args-and-log-builder/22695
- https://stackoverflow.com/questions/76157524/temporary-value-dropped-while-borrowed-when-printing-format-args
- https://internals.rust-lang.org/t/format-args-with-long-lifetimes/19494

I understand if the `format_args!` macro cannot be changed, but then please provide an alternative way of building `fmt::Arguments` without that restriction. Even a small performance overhead (for example cloning the values) would be an improvement compared to the common workarounds that have to be used otherwise.

**Comment by HindrikStegenga** [comment]
So I ran into this issue, but at the same time I managed to find a rather nifty workaround.
Basically it works if you use an anonymous closure and directly invoke it. (In my case I needed the length of the outputted string)
```
        (|buf : &mut fmt::Formatter, args : std::fmt::Arguments<'_>| {
            let i = args.to_string().len();
            buf.write_fmt(args)?;

            for _ in 0..(128 - i) {
                write!(buf, " ")?;
            }
            writeln!(buf, "{}:{}", file, line)
        })(buf, format_args!("[{}][{}] - {}", level, target, args))
```
For me this works and compiles on stable 1.61.

**Comment by piegamesde** [comment]
Yes, this looks pretty much like the same hack fern is using with its `FormatCallback`.

**Comment by albel727** [comment]
@HindrikStegenga 
> (In my case I needed the length of the outputted string)

An exercise in futility, since you're calling `args.to_string()`, i.e. allocating a temporary string with the output anyway. You could've just called `format!()` right away and reused that same string for the rest of your work with `buf.write_str()`, instead of bothering with `format_args!()` and then redoing the formatting work a second time with `buf.write_fmt()`.

If you really want to get the length of the outputted string without allocating, you could try calculating it manually, e.g. `level.len() + target.len() + args.len() + 7`. Or, more generally, you could make your implementor of `std::fmt::Write`, that would sum the lengths of the received strings in its `write_str()`. You could then feed your args to it with `std::fmt::write()`.

[Example in Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=564839ff291559e49eaf93d327cb6f83)

**Comment by tmccombs** [comment]
At the very least this surprising behavior should be better documented. And maybe, the error message should give a better recommendation of what to do when this happens for `format_args!`. 

The suggestion of changing `let args = format_args!("{}", thingJ);` to 
```rust
let binding = format_args!("{}", lazy);
let args = binding;
```

isn't really helpful.

**Comment by piegamesde** [comment]
With the changes made to `format_args!` in https://github.com/rust-lang/rust/pull/106745/, would it allow us to fix this issue?

**Comment by m-ou-se** [comment]
> With the changes made to `format_args!` in #106745, would it allow us to fix this issue?

Nope, that's unrelated. format_args's expansion includes temporaries whose lifetimes need to be extended. We currently don't have a flexible way to do that. Temporary lifetime extension only applies to very few types of expressions, which don't include function calls. See also some related thoughts here on zulip: https://rust-lang.zulipchat.com/#narrow/stream/213817-t-lang/topic/design.20meeting.202023-03-15.3A.20temporary.20lifetimes/near/351415051

**Comment by Stargateur** [comment]
> At the very least this surprising behavior should be better documented. And maybe, the error message should give a better recommendation of what to do when this happens for `format_args!`.
> 
> The suggestion of changing `let args = format_args!("{}", thing);` to
> 
> ```rust
> let binding = format_args!("{}", thing);
> let args = binding;
> ```
> 
> isn't really helpful.

Yes the error message is very misleading and doesn't make sense. We can add [this ](https://stackoverflow.com/q/76157524/7076153) stackoverflow question to the list of confused people.

**Comment by daniel-pfeiffer** [comment]
To make a deeply nested `format_args` more readable, I want to pull it apart
```rust
let y = format_args!("<{x}>");
println!("{y}");
```
As noted above, this gives
```
6 | let y = format_args!("<{x}>");
  |         ^^^^^^^^^^^^^^^^^^^^^- temporary value is freed at the end of this statement
  |         |
  |         creates a temporary value which is freed while still in use
7 | println!("{y}");
  |           --- borrow later used here
  |
  = note: this error originates in the macro `format_args` (in Nightly builds, run with -Z macro-backtrace for more info)
help: consider using a `let` binding to create a longer lived value
  |
6 + let binding = format_args!("<{x}>");
7 ~ let y = binding;
```

Weird suggestion, but if I follow it
```rust
let binding = format_args!("<{x}>");
let y = binding;
println!("{y}");
```
It chokes on its own previous suggestion, giving an even weirder one, as it doesn't see `binding` is already used
```
6 | let binding = format_args!("<{x}>");
  |               ^^^^^^^^^^^^^^^^^^^^^- temporary value is freed at the end of this statement
  |               |
  |               creates a temporary value which is freed while still in use
7 | let y = binding;
  |         ------- borrow later used here
  |
  = note: this error originates in the macro `format_args` (in Nightly builds, run with -Z macro-backtrace for more info)
help: consider using a `let` binding to create a longer lived value
  |
6 + let binding = format_args!("<{x}>");
7 ~ let binding = binding;
```

So I thought: if some temporary is too short lived, let me turn it all into one statement. Essentially like @HindrikStegenga's passing it to a closure, but in natural order
```rust
if let y = format_args!("<{x}>") {
    println!("{y}");
}
```
Even though this works as desired, it gives
```
6 | if let y = format_args!("<{x}>") {
  |    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: this pattern will always match, so the `if let` is useless
  = help: consider replacing the `if let` with a `let`
  = note: `#[warn(irrefutable_let_patterns)]` on by default
```

**Comment by daniel-pfeiffer** [comment]
As this doesn't seem to be moving fast, I wrapped @HindrikStegenga's closure in natural syntax:

```rust
macro_rules! let_workaround {
    (let $name:ident = $val:expr; $($rest:tt)+) => {
        (|$name| { let_workaround!{$($rest)+} })($val)
    };
    ($($rest:tt)+) => { $($rest)+ }
}

fn main() {
    let (a, b, c) = (1, 2, 3);
    let fab = let_workaround! {
        let fa = format_args!("a {a}");
        let fb = format_args!("b {b}");
        format!("{fa}, {fb}") // return value needs semicolon on outer statement
    };
    let_workaround! {
        let fc = format_args!("c {c}");
        let fabc = format_args!("{fab}, {fc}");
        println!("{fabc}");
    } // no semicolon on block
    println!("done");
}
```

What this still doesn't solve is the hindrance caused by the lack of `? :`. `if else` just isn't equivalent, as `format_args!` also can't be returned from a block. E.g. when you need to choose formatting at runtime:

```rust
if something { format_args!("a {a}") }
else { format_args!("b {b}") }
```

**Comment by daniel-pfeiffer** [comment]
Turns out [clippy doesn't like this](https://rust-lang.github.io/rust-clippy/master/index.html#/redundant_closure_call). What clippy misses here, is that this currently seems to be the only way to extend the temp lifetimes. So clippy is wrong. :-1:

**Comment by dtolnay** [comment]
Use `match`:

```rust
macro_rules! let_workaround {
    (let $name:ident = $val:expr; $($rest:tt)+) => {
        match $val {
            $name => {
                let_workaround! { $($rest)+ }
            }
        }
    };
    ($($rest:tt)+) => { $($rest)+ }
}
```

**Comment by quaternic** [comment]
@daniel-pfeiffer 
> What this can't solve, is the lack of `? :`. `if else` just isn't equivalent, as `format_args!` also can't be returned from a block. E.g. you can't choose efficient low-level formatting at runtime:
> 
> ```rust
> if something { format_args!("a {a}") }
> else { format_args!("b {b}") }
> ```
There is a workaround for that:
```rust
macro_rules! select {
    ($cond:expr, $iftrue:expr, $iffalse:expr) => {
        'outer:  {
            (
                'inner: {
                    if $cond { break 'inner }
                    break 'outer $iffalse
                },
                $iftrue
            ).1
        }
    }
}
```
To clarify the structure, that's essentially like
```rust
let tuple = (
    if condition { break iftrue } ,
    iffalse
);
break tuple.1;
```
or just
```rust
if condition {
    break iftrue
}
break iffalse
```

so `select!(cond, A, B)` means the same as `if cond { A } else { B }` but without the separate drop scopes for the branches.

In combination with the above variant of `let_workaround` from @dtolnay, you can write stuff like
```rust
let_workaround! {
    let ab = select!(a < b, format_args!(" < {b}"),
             select!(a > b, format_args!(" > {b}"),
                            format_args!("")));
    let bc = select!(b < c, format_args!(" < {c}"),
             select!(b > c, format_args!(" > {c}"),
                            format_args!("")));
    let ca = select!(c < a, format_args!(" < {a}"),
             select!(c > a, format_args!(" > {a}"),
                            format_args!("")));
    let min = a.min(b).min(c);
    let f_abc = select!(min == a, format_args!("{a}{ab}{bc}{ca}"),
                select!(min == b, format_args!("{b}{bc}{ca}{ab}"),
                select!(min == c, format_args!("{c}{ca}{ab}{bc}"),
                                  unreachable!())));
    write!(f, "{f_abc}")
}
```
https://godbolt.org/z/EcKqKTKGq

(This came up in related discussion in https://internals.rust-lang.org/t/format-args-with-long-lifetimes/19494)

**Comment by justinhaubrich** [comment]
Has this issue been resolved yet? Kind of crazy that it doesn't work...

**Comment by justinhaubrich** [comment]
> So I ran into this issue, but at the same time I managed to find a rather nifty workaround. Basically it works if you use an anonymous closure and directly invoke it. (In my case I needed the length of the outputted string)
> 
> ```
>         (|buf : &mut fmt::Formatter, args : std::fmt::Arguments<'_>| {
>             let i = args.to_string().len();
>             buf.write_fmt(args)?;
> 
>             for _ in 0..(128 - i) {
>                 write!(buf, " ")?;
>             }
>             writeln!(buf, "{}:{}", file, line)
>         })(buf, format_args!("[{}][{}] - {}", level, target, args))
> ```
> 
> For me this works and compiles on stable 1.61.

Thanks for this. Using a closure is the only way I was able to get it to work.

**Comment by daniel-pfeiffer** [comment]
Since Rust hasn’t yet solved these problems, I’ve extended these solutions and [published them](https://crates.io/crates/temporaries).

**Comment by ArhanChaudhary** [comment]
Update: this will hopefully be fixed by https://github.com/rust-lang/rust/pull/139080

**Comment by m-ou-se** [comment]
> Update: this will hopefully be fixed by [#139080](https://github.com/rust-lang/rust/pull/139080)

Indeed!

Here's the solution for this issue: https://github.com/rust-lang/rust/pull/139135

(Blocked on a few other PRs, and blocked on an FCP of course.)

**Comment by NobodyXu** [comment]
I think it would need super let to work?
